dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeerContext.cpp

```diff
@@ -209,4 +222,63 @@ void FastAddPeerContext::insertCheckpointIngestInfo(
     info->persistToLocal();
 }
 
+void FastAddPeerContext::resolveFapSnapshotState(
+    TMTContext & tmt,
+    const TiFlashRaftProxyHelper * proxy_helper,
+    UInt64 region_id,
+    bool is_legacy_snapshot)
+{
+    auto prev_state = tasks_trace->asyncCancelTask(
+        region_id,
+        [&]() {},
+        false);
+    // Legacy snapshot and FAP(both phase 1 and 2) for a region is exclusive for now.
+    if (prev_state == FAPAsyncTasks::TaskState::Finished)
+    {
+        if (is_legacy_snapshot)
+        {
+            // Proxy will actively polling FAP result by calling `fn_fast_add_peer`,
+            // so the result will be eventually fetched by Proxy, and then trigger phase2.
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "FastAddPeer: find unfetched finished fap task, make sure proxy has longer timeout threshold, "
+                "region_id={}",
+                region_id);
+        }
+        else
+        {
+            LOG_INFO(log, "FastAddPeer: Cancel finished FAP because region destroyed, region_id={}", region_id);
+            cleanTask(tmt, proxy_helper, region_id, false);
+        }
+    }
+    else if likely (prev_state == FAPAsyncTasks::TaskState::NotScheduled)
+    {
+        // 1. There leaves some non-ingested data on disk after restart.
+        // 2. There has been no fap at all.
+        // 3. FAP is enabled before, but disabled for now.
+        LOG_DEBUG(log, "FastAddPeer: no find ongoing fap task, region_id={}", region_id);
+        // Still need to clean because there could be data left.
+        cleanTask(tmt, proxy_helper, region_id, false);
+    }
+    else
+    {
+        if (is_legacy_snapshot)
+        {
+            // Currently, proxy will not actively cancel FAP. So it will not fallback if FAP phase 1 is still running.
+            // If we cancel the task asyncly, this will happen in some very corner cases where a legacy snapshot meets a cleaning FAP task.
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "FastAddPeer: find running scheduled fap task, region_id={} fap_state={}",
+                region_id,
+                magic_enum::enum_name(prev_state));
+        }
+        else
+        {
+            // If the region is sent to this store later, it will be with another peer_id.
+            LOG_INFO(log, "FastAddPeer: Cancel running/queueing FAP because region destroyed, region_id={}", region_id);
+            cleanTask(tmt, proxy_helper, region_id, false);
```

Comment:
We'd to block wait the FAP to cancel to prevent potential.
There is no race for now, since either `handleDestroy` or `cleanTask` handles their own data. However, it is not a good way.